### PR TITLE
LS-8690: BE changes for allowing non unique batches for material with…

### PR DESCRIPTION
**Pull request by iostream04**
… different manufacturer

### CODE REVIEW

File: modules/api/src/main/java/com/logistimo/api/models/DemandItemBatchModel.java [hunk 1]
```diff
@@ -32,7 +32,7 @@
 /**
  * @author charan
  */
-public class DemandItemBatchModel {
+public class  DemandItemBatchModel {
```
Comment (surajk94) [inline modules/api/src/main/java/com/logistimo/api/models/DemandItemBatchModel.java:35]: revert change

File: modules/api/src/main/java/com/logistimo/api/servlets/mobile/builders/MobileShipmentBuilder.java [hunk 2]
```diff
@@ -454,8 +457,9 @@ MobileShipmentItemBatchModel buildMobileShipmentItemBatchModel(IShipmentItemBatc
 
       if (iAllocs != null && !iAllocs.isEmpty()) {
         for (IInvAllocation iAlloc : iAllocs) {
-          if (iAlloc.getBatchId() != null && !iAlloc.getBatchId().isEmpty() && iAlloc.getBatchId()
-              .equals(sib.getBatchId())) {
+          if (iAlloc.getBatchId() != null && !iAlloc.getBatchId().isEmpty()
```
Comment (surajk94) [inline modules/api/src/main/java/com/logistimo/api/servlets/mobile/builders/MobileShipmentBuilder.java:460]: use StringUtils.isNotBlank() instead

File: modules/api/src/main/java/com/logistimo/api/servlets/mobile/builders/MobileShipmentBuilder.java [hunk 3]
```diff
@@ -454,8 +457,9 @@ MobileShipmentItemBatchModel buildMobileShipmentItemBatchModel(IShipmentItemBatc
...
+                  && iAlloc.getBatchId().equals(sib.getBatchId()) && iAlloc.getCatalogueId()!= null
```
Comment (surajk94) [inline modules/api/src/main/java/com/logistimo/api/servlets/mobile/builders/MobileShipmentBuilder.java:461]: use StringUtils.isNotBlank() instead

File: modules/api/src/test/java/com/logistimo/api/builders/InventoryBuilderTest.java [hunk 4]
```diff
@@ -104,9 +109,10 @@ public void testSetBatchModel() throws ServiceException {
     invntryBatches.add(setInvntryBatch("B002", "Serum Institute of Medicines", new BigDecimal(100), new Date(), new Date(1599806451000l)));
 
     when(inventoryManagementService.getBatches(any(), any(), any())).thenReturn(new Results<>(invntryBatches, null));
+    when(manufacturerCatalogueService.getManufacturerNameFromCatalogueId(Mockito.anyString()))
+            .thenReturn("Serum Institute").thenReturn("Serum Institute of Medicines");
```
Comment (surajk94) [inline modules/api/src/test/java/com/logistimo/api/builders/InventoryBuilderTest.java:113]: why double "thenReturn"?

File: modules/inventory/src/main/java/com/logistimo/inventory/dao/impl/InvntryDao.java [hunk 5]
```diff
@@ -257,10 +257,10 @@ public IInvntryBatch findInvBatch(Long kioskId, Long materialId, String batchId,
         q =
         pm.newQuery("SELECT FROM " + JDOUtils.getImplClassName(IInvntryBatch.class)
             + " WHERE kId == kIdParam && " +
-            "mId == mIdParam && bid == bIdParam PARAMETERS Long kIdParam, Long mIdParam, String bIdParam");
+            "CATALOGUE_ID == catalogueIdParam && bid == bIdParam PARAMETERS Long kIdParam, Long mIdParam, String bIdParam");
```
Comment (surajk94) [inline modules/inventory/src/main/java/com/logistimo/inventory/dao/impl/InvntryDao.java:260]: CATALOGUE_ID should be the variable name in IInvntryBatch class

File: modules/inventory/src/main/java/com/logistimo/inventory/entity/IInvntryLog.java [hunk 6]
```diff
@@ -66,4 +66,6 @@
   String getArchivedBy();
 
   void setArchivedBy(String archivedBy);
+
```
Comment (surajk94) [inline modules/inventory/src/main/java/com/logistimo/inventory/entity/IInvntryLog.java:69]: revert changes

File: modules/inventory/src/main/java/com/logistimo/inventory/service/InventoryManagementService.java [hunk 7]
```diff
@@ -148,7 +148,7 @@ Results getInventoryByBatchExpiry(Long domainId, Long materialId, Date start, Da
   /**
    * Get valid batch for a given batchId, materialId and storeId.
```
Comment (surajk94) [inline modules/inventory/src/main/java/com/logistimo/inventory/service/InventoryManagementService.java:149]: update java doc with catalogueId instead of materialId

File: modules/orders/src/main/java/com/logistimo/orders/models/InvoiceItem.java [hunk 8]
```diff
@@ -150,7 +156,7 @@ public void setBatchEnabled(Boolean batchEnabled) {
   }
 
   public String getBatch() {
-    if (batchId != null) {
+    if (batchId != null) { // need of catalogueId ??
```
Comment (surajk94) [inline modules/orders/src/main/java/com/logistimo/orders/models/InvoiceItem.java:159]: dont think catalogue id is needed, as manufacturer name is already being used. recheck once

File: modules/orders/src/main/java/com/logistimo/orders/service/impl/DemandService.java [hunk 9]
```diff
@@ -808,10 +815,11 @@ public String getMaterialStatusForDemandItem(String id, Long oId, Long mId) {
 
       if (iAllocs != null && !iAllocs.isEmpty()) {
         for (IInvAllocation iAlloc : iAllocs) {
-          if (iAlloc.getBatchId() != null && !iAlloc.getBatchId().isEmpty()) {
+          if (iAlloc.getBatchId() != null && !iAlloc.getBatchId().isEmpty() && iAlloc.getCatalogueId() != null && !iAlloc.getCatalogueId().isEmpty()) {
```
Comment (surajk94) [inline modules/orders/src/main/java/com/logistimo/orders/service/impl/DemandService.java:818]: use StringUtils.isNotBlank

File: modules/orders/src/test/java/com/logistimo/orders/actions/OrderAutomationActionTest.java [hunk 10]
```diff
@@ -38,10 +38,13 @@
 import com.logistimo.orders.service.impl.OrderManagementServiceImpl;
 import com.logistimo.pagination.Results;
 import com.logistimo.services.ServiceException;
+import com.logistimo.shipments.action.UpdateShipmentStatusAction;
 import com.logistimo.utils.BigUtil;
 
 import org.junit.Before;
 import org.junit.Test;
+import org.mockito.InjectMocks;
```
Comment (surajk94) [inline modules/orders/src/test/java/com/logistimo/orders/actions/OrderAutomationActionTest.java:46]: are these additional imports needed? no change in code...

File: modules/returns/src/main/java/com/logistimo/returns/models/ReturnsItemBatchModel.java [hunk 11]
```diff
@@ -44,7 +44,7 @@
 
   private String manufacturer;
 
-  @SerializedName("catalogue_id")
+  @SerializedName("CATALOGUE_ID")
```
Comment (surajk94) [inline modules/returns/src/main/java/com/logistimo/returns/models/ReturnsItemBatchModel.java:47]: revert this change

File: modules/returns/src/test/java/com/logistimo/returns/actions/UpdateReturnActionTest.java [hunk 12]
```diff
@@ -81,6 +81,7 @@ public void testUpdateReturns() throws IOException {
     when(getReturnsAction.invoke(any())).thenReturn(oldReturnVO);
     when(returnsRepository.saveReturnsTrackingDetails(any())).thenReturn(updatedTrackingVO);
     ReturnsVO updatedReturnsVO = updateReturnAction.invoke(newReturnVO);
+
```
Comment (surajk94) [inline modules/returns/src/test/java/com/logistimo/returns/actions/UpdateReturnActionTest.java:84]: revert change

File: modules/api/src/main/java/com/logistimo/api/servlets/mobile/InventoryServlet.java [hunk 13]
```diff
@@ -860,6 +856,16 @@ private void validateMobileUpdateInvTransRequest(MobileUpdateInvTransRequest mob
     if (StringUtils.isEmpty(mobUpdateInvTransReq.uid) || mobUpdateInvTransReq.kid == null || mobUpdateInvTransReq.sntm == null || mobUpdateInvTransReq.trns == null || mobUpdateInvTransReq.trns.isEmpty()) {
       throw new InvalidDataException(backendMessages.getString("error.invaliddata.frommobile"));
     }
+    for(MobileMaterialTransModel transaction : mobUpdateInvTransReq.trns){
+      if(transaction.trns != null && !transaction.trns.isEmpty()) {
+        for (MobileTransModel materialTransaction : transaction.trns) {
+          if(materialTransaction.bid != null && !materialTransaction.bid.isEmpty()
```
Comment (surajk94) [inline modules/api/src/main/java/com/logistimo/api/servlets/mobile/InventoryServlet.java:862]: use Stringutils.isnotblank()

File: modules/api/src/main/java/com/logistimo/api/servlets/mobile/InventoryServlet.java [hunk 14]
```diff
@@ -860,6 +856,16 @@ private void validateMobileUpdateInvTransRequest(MobileUpdateInvTransRequest mob
...
+                  && (materialTransaction.catalogueId == null || materialTransaction.catalogueId.isEmpty())){
+            throw new InvalidDataException(backendMessages.getString("error.invaliddata.frommobile"));
```
Comment (surajk94) [inline modules/api/src/main/java/com/logistimo/api/servlets/mobile/InventoryServlet.java:864]: Is this check needed here? Is it not handled in service layer?
Only batch materials will have batchid and catalogueid. remaining material will have these values as null
Review if this change is needed

File: modules/api/src/test/java/com/logistimo/api/builders/InventoryBuilderTest.java (hunk 4 above)
Comment (iostream04) [reply modules/api/src/test/java/com/logistimo/api/builders/InventoryBuilderTest.java:113]: First call it will return the first parameter provided, On second call it will return the second parameter provided. Different Manufacturer Names to validate.

File: modules/inventory/src/main/java/com/logistimo/inventory/dao/impl/InvntryDao.java (hunk 5 above)
Comment (iostream04) [reply modules/inventory/src/main/java/com/logistimo/inventory/dao/impl/InvntryDao.java]: Done!

File: modules/orders/src/main/java/com/logistimo/orders/models/InvoiceItem.java (hunk 8 above)
Comment (iostream04) [reply modules/orders/src/main/java/com/logistimo/orders/models/InvoiceItem.java:159]: yes, No need of catalogueId,I will remove it.

File: modules/api/src/main/java/com/logistimo/api/builders/ReturnsBuilder.java [hunk 15]
```diff
@@ -336,13 +336,13 @@ private ReturnsItemBatchVO getReturnsItemBatchVO(Long vendorId, Long customerId,
...
-        .getInventoryBatch(vendorId, materialId, itemBatchModel.getBatchId(), null);
+        .getInventoryBatch(vendorId, itemBatchModel.getCatalogueId(), itemBatchModel.getBatchId(), null);
 
     if (inventoryBatch == null) {
       inventoryBatch = inventoryManagementService
-          .getInventoryBatch(customerId, materialId, itemBatchModel.getBatchId(), null);
+          .getInventoryBatch(customerId, itemBatchModel.getCatalogueId(), itemBatchModel.getBatchId(),null);
       if (inventoryBatch == null) {
-        return new BatchVO(itemBatchModel.getBatchId(), null, null, null, null);
+        return new BatchVO(itemBatchModel.getBatchId(), null, null, itemBatchModel.getCatalogueId(), null);
```
[the remark this answers is not in the document]
Comment (iostream04) [reply modules/api/src/main/java/com/logistimo/api/builders/ReturnsBuilder.java:345]: Done!

File: modules/api/src/main/java/com/logistimo/api/servlets/mobile/InventoryServlet.java (hunk 14 above)
Comment (iostream04) [reply modules/api/src/main/java/com/logistimo/api/servlets/mobile/InventoryServlet.java:864]: It was just cross check if mobile API is sending the catalogueId details or not. But no needed.

File: modules/api/src/main/java/com/logistimo/api/servlets/mobile/builders/MobileShipmentBuilder.java (hunk 2 above)
Comment (iostream04) [reply modules/api/src/main/java/com/logistimo/api/servlets/mobile/builders/MobileShipmentBuilder.java:460]: Added!

File: modules/api/src/main/java/com/logistimo/api/servlets/mobile/builders/MobileShipmentBuilder.java (hunk 3 above)
Comment (iostream04) [reply modules/api/src/main/java/com/logistimo/api/servlets/mobile/builders/MobileShipmentBuilder.java:461]: Added !

File: modules/inventory/src/main/java/com/logistimo/inventory/entity/IInvntryLog.java (hunk 6 above)
Comment (iostream04) [reply modules/inventory/src/main/java/com/logistimo/inventory/entity/IInvntryLog.java:69]: Done!

File: modules/orders/src/main/java/com/logistimo/orders/service/impl/DemandService.java (hunk 9 above)
Comment (iostream04) [reply modules/orders/src/main/java/com/logistimo/orders/service/impl/DemandService.java:818]: Added!

File: modules/orders/src/test/java/com/logistimo/orders/actions/OrderAutomationActionTest.java (hunk 10 above)
Comment (iostream04) [reply modules/orders/src/test/java/com/logistimo/orders/actions/OrderAutomationActionTest.java:46]: Removed!

File: modules/returns/src/main/java/com/logistimo/returns/models/ReturnsItemBatchModel.java (hunk 11 above)
Comment (iostream04) [reply modules/returns/src/main/java/com/logistimo/returns/models/ReturnsItemBatchModel.java:47]: Done!

File: modules/api/src/main/java/com/logistimo/api/models/DemandItemBatchModel.java (hunk 1 above)
Comment (iostream04) [reply modules/api/src/main/java/com/logistimo/api/models/DemandItemBatchModel.java]: Done!

File: modules/api/src/main/java/com/logistimo/api/builders/DemandBuilder.java [hunk 16]
```diff
@@ -227,7 +228,8 @@ public DemandModel buildDemandModel(Long kioskId, Long materialId,
         breakdownModel.orderId = model.oid;
         breakdownModel.aQty = ib.getAllocatedStock();
         for (IInvAllocation invAllocation : inv) {
-          if (breakdownModel.bid.equals(invAllocation.getBatchId())) {
+          if ((breakdownModel.bid + "\t" + breakdownModel.catalogueId)
```
[the remark this answers is not in the document]
Comment (iostream04) [reply modules/api/src/main/java/com/logistimo/api/builders/DemandBuilder.java:231]: Done!

File: modules/api/src/main/java/com/logistimo/api/servlets/mobile/InventoryServlet.java (hunk 13 above)
Comment (iostream04) [reply modules/api/src/main/java/com/logistimo/api/servlets/mobile/InventoryServlet.java]: Done!

File: modules/api/src/main/java/com/logistimo/api/controllers/TransactionsController.java [hunk 17]
```diff
@@ -583,38 +584,51 @@ String addTransaction(@RequestBody Map<String, Object> transaction) {
...
-          String status = String.valueOf(mat.get("mst"));
-          if (status.equals("null")) {
-            status = "";
-          }
-          if (checkBatchMgmt) {
-            IMaterial material = materialCatalogService.getMaterial(materialId);
-            if (material.isBatchEnabled()) {
-              berrorMaterials.add(material.getName());
-            }
+      for (Map<String,Object> item : items) {
```
Comment (charanm) [inline modules/api/src/main/java/com/logistimo/api/controllers/TransactionsController.java:589]: I thought we discussed to refactor this API to a simpler model. Is it not possible to defined a model object and use "@RequestBody" directly?

File: modules/api/src/main/java/com/logistimo/api/controllers/TransactionsController.java [hunk 18]
```diff
@@ -583,38 +584,51 @@ String addTransaction(@RequestBody Map<String, Object> transaction) {
                 .parseCustom(String.valueOf(transaction.get("transactual")), Constants.DATE_FORMAT,
                     null);
       }
-      List<LinkedTreeMap> items = (List) transaction.get("materials");
+      List<Map<String, Object>> items = (List) transaction.get("materials");
```
Comment (charanm) [inline modules/api/src/main/java/com/logistimo/api/controllers/TransactionsController.java:587]: LinkedTreeMap assumes the sequence should be maintained, but switching to Map does not guarantee it right/

File: modules/api/src/main/java/com/logistimo/api/controllers/TransactionsController.java [hunk 19]
```diff
@@ -583,38 +584,51 @@ String addTransaction(@RequestBody Map<String, Object> transaction) {
...
-            status = "";
-          }
-          if (checkBatchMgmt) {
-            IMaterial material = materialCatalogService.getMaterial(materialId);
-            if (material.isBatchEnabled()) {
-              berrorMaterials.add(material.getName());
-            }
+      for (Map<String,Object> item : items) {
+
+        Long materialId = (long)Double.parseDouble(String.valueOf(item.get("mid")));
```
Comment (charanm) [inline modules/api/src/main/java/com/logistimo/api/controllers/TransactionsController.java:591]: Why this unnecessary Double parsing? The server should never accept anything other than a long.

File: modules/api/src/main/java/com/logistimo/api/controllers/TransactionsController.java [hunk 20]
```diff
@@ -583,38 +584,51 @@ String addTransaction(@RequestBody Map<String, Object> transaction) {
...
+        String reason = String.valueOf(item.get("r"));
+        if (reason.equals("null")) {
+          reason = "";
+        }
+        String status = String.valueOf(item.get("mst"));
+        if (status.equals("null")) {
+          status = "";
+        }
+        boolean hasBatch = validateBatchDetails(item);
+        if (checkBatchMgmt) {
```
Comment (charanm) [inline modules/api/src/main/java/com/logistimo/api/controllers/TransactionsController.java:603]: This code block is confusing. we validateBatchDetails to set hasBatch, and follow it by checking batch management.

File: modules/api/src/main/java/com/logistimo/api/controllers/TransactionsController.java [hunk 21]
```diff
@@ -583,38 +584,51 @@ String addTransaction(@RequestBody Map<String, Object> transaction) {
...
-            if (material.isBatchEnabled()) {
-              berrorMaterials.add(material.getName());
-            }
+      for (Map<String,Object> item : items) {
+
+        Long materialId = (long)Double.parseDouble(String.valueOf(item.get("mid")));
+
+        BigDecimal quantity = new BigDecimal(Long.parseLong(String.valueOf(item.get("q"))));
+        String reason = String.valueOf(item.get("r"));
+        if (reason.equals("null")) {
```
Comment (charanm) [inline modules/api/src/main/java/com/logistimo/api/controllers/TransactionsController.java:595]: Plumbing work related to "null" seems to come from the webapp, this should be handled there.

File: modules/api/src/main/java/com/logistimo/api/controllers/TransactionsController.java [hunk 22]
```diff
@@ -583,38 +584,51 @@ String addTransaction(@RequestBody Map<String, Object> transaction) {
...
+          if (manufactured.equals("null")) {
+            manufactured = "";
           }
-          ITransaction
-              trans =
-              getTransaction(userId, domainId, transType, kioskId, linkedKioskId, reason, status,
-                  now,
-                  materialId, quantity, "", actualTransDate, trkid);
-          transList.add(trans);
+          trans = getTransaction(userId, domainId, transType, kioskId, linkedKioskId, reason, status,
```
Comment (charanm) [inline modules/api/src/main/java/com/logistimo/api/controllers/TransactionsController.java:624]: There are too many arguments, define a TransactionBuilder.

File: modules/api/src/main/java/com/logistimo/api/controllers/TransactionsController.java [hunk 23]
```diff
@@ -583,38 +584,51 @@ String addTransaction(@RequestBody Map<String, Object> transaction) {
...
+                  now, materialId, quantity, batchId, actualTransDate, trkid);
+          TransactionUtil.setBatchData(trans, batchId, expiry, manufacturer, catalogueId, manufactured);
```
Comment (charanm) [inline modules/api/src/main/java/com/logistimo/api/controllers/TransactionsController.java:626]: Once builder is defined, there is no reason to have two separate steps, this can be merged into it.

File: modules/api/src/main/java/com/logistimo/api/controllers/TransactionsController.java [hunk 24]
```diff
@@ -700,6 +675,18 @@ String addTransaction(@RequestBody Map<String, Object> transaction) {
...
   }
+  private boolean validateBatchDetails(Map<String, Object> item){
+    try{
+      String batchId = String.valueOf(item.get("bid"));
+      String catalogueId = String.valueOf(item.get("catalogue_id"));
+      if(batchId.equals("null") || catalogueId.equals("null")){
+        return false;
+      }
+    }catch (Exception e){
+      return false;
```
Comment (charanm) [inline modules/api/src/main/java/com/logistimo/api/controllers/TransactionsController.java:686]: What exception can happen here, that needs to be suppressed. batchId and catalogueId can be null, should be handled gracefully.

File: modules/api/src/main/java/com/logistimo/api/servlets/mobile/builders/MobileShipmentBuilder.java [hunk 25]
```diff
@@ -454,8 +457,8 @@ MobileShipmentItemBatchModel buildMobileShipmentItemBatchModel(IShipmentItemBatc
 
       if (iAllocs != null && !iAllocs.isEmpty()) {
         for (IInvAllocation iAlloc : iAllocs) {
-          if (iAlloc.getBatchId() != null && !iAlloc.getBatchId().isEmpty() && iAlloc.getBatchId()
-              .equals(sib.getBatchId())) {
+          if (StringUtils.isNotBlank(iAlloc.getBatchId()) && iAlloc.getBatchId().equals(sib.getBatchId())
```
Comment (charanm) [inline modules/api/src/main/java/com/logistimo/api/servlets/mobile/builders/MobileShipmentBuilder.java:460]: This is the point with using ValueObjects, we ended up repeating these conditions everywhere in the code.

File: modules/common/src/main/java/com/logistimo/proto/BatchRequest.java [hunk 26]
```diff
@@ -42,4 +42,8 @@
    * Material status
    */
   public String mst;
+  /**
+   * catalogueId
```
Comment (charanm) [inline modules/common/src/main/java/com/logistimo/proto/BatchRequest.java:46]: Can you add something more meaningful than repeating the variable name?

File: modules/event-handlers/src/main/java/com/logistimo/events/templates/InvntryBatchTemplate.java [hunk 27]
```diff
@@ -90,6 +90,8 @@ public InvntryBatchTemplate(IInvntryBatch invntryBatch) {
       // Batch
       if (excludeVars == null || !excludeVars.contains(EventsConfig.VAR_BATCHNUMBER)) {
         varMap.put(EventsConfig.VAR_BATCHNUMBER, invntryBatch.getBatchId());
+       // catalogueId needed here ?
```
Comment (charanm) [inline modules/event-handlers/src/main/java/com/logistimo/events/templates/InvntryBatchTemplate.java:93]: Good question. These values are used while building Notifications(SMS/Email..). Batch independently does not mean the same anymore, this should always be qualified with Manufacturer. You can take this up with PMs, to see if  we should always include Manufacturer name along with Batch number, in a format like "BX1234(Bio tech)".

**Comment by charanm** [comment]
jarvis

File: modules/event-handlers/src/main/java/com/logistimo/events/templates/InvntryBatchTemplate.java (hunk 27 above)
Comment (iostream04) [reply modules/event-handlers/src/main/java/com/logistimo/events/templates/InvntryBatchTemplate.java:93]: After taking PMs input we would we including manufacturer name in following pattern.
manufacturer names can be long, maybe we can restrict or ellipsize the manu. name If it is more than 10 character; 
example: BX1234 (Biotech), or BX1234 (Serum Inst...).
